### PR TITLE
added monoid action on functional spaces.

### PR DESCRIPTION
**Pull request by piyush-kurur**
Let s be any type and let m act on it. Then one can lift the action of the monoid m to the functional space
`s -> a` for any type `a`. This is essentially the point wise multiplication.  If in addition a is a monoid, then
`s -> a` is a monoid  and lifted action is nice in the sense that it satisfies the additional laws required for monoid action on monoids. I guess therefore this instance is a natural addition.

### CODE REVIEW

**Comment by byorgey** [comment]
This makes me uneasy, for two reasons.  One, there is another natural instance
```
instance Action m a => Action m (s -> a) where
  act m f s = m `act` f s
```
These instances overlap and I don't see any reason to prefer one over the other.

Secondly, although I know the instance you give is the right one in the case of pointer manipulations, there are other cases where you really want m to be a group and to have it act inversely, to match with the contravariance of `s`.  Something like this:
```
instance (Group m, Action m s) => Action m (s -> a) where
  act m f s = f $ inverse m `act` s
```
So, in short, there are a lot of choices here and I don't see good reasons to prefer one choice in particular.

**Comment by piyush-kurur** [comment]
On Sat, Dec 05, 2015 at 03:28:16PM -0800, Brent Yorgey wrote:
> This makes me uneasy, for two reasons.  One, there is another natural instance
> ```
> instance Action m a => Action m (s -> a) where
>   act m f s = m `act` f s
> ```
> These instances overlap and I don't see any reason to prefer one over the other.
> 
> Secondly, although I know the instance you give is the right one in the case of pointer manipulations, there are other cases where you really want m to be a group and to have it act inversely, to match with the contravariance of `s`.  Something like this:
> ```
> instance (Group m, Action m s) => Action m (s -> a) where
>   act m f s = f $ inverse m `act` s
> ```
> So, in short, there are a lot of choices here and I don't see good reasons to prefer one choice in particular.
> 

Agree with you. Does it make sense to distinguish between right and
left actions and make one action canonical for left and another for
right.

**Comment by byorgey** [comment]
Is there any canonical mathematical reason to associate one particular lifting with left actions and the other with right actions?  If not, I would say no.  I want to avoid making arbitrary choices.

**Comment by piyush-kurur** [comment]
I cannot justify either of this actions so I think this pull request
can be closed without a merge. However, I would like to "outsource" as
much as possible, the monoid actions from the raaz package to this. Do
you have specific ideas for this. Yes there is always the new type
wrapping and defining it locally but since function application is a
basic type, I just wanted to make sure that I do not miss something
out.

**Comment by byorgey** [comment]
I mean, I would be quite happy to define some `newtype`s with these instances in the `monoid-extras` package.  There is no reason newtypes have to be done locally.

**Comment by piyush-kurur** [comment]
I am undecided on whether it is really worth doing this.  For
applications in our library, a further generalisations to arrows is
required; in fact for the applicative functor `WrappedArrow` See the
last line of

<https://github.com/raaz-crypto/raaz/blob/master/Raaz/Core/MonoidalAction.hs>

This generalisation avoids unnecessary code duplication across pure
and impure IO function. But then we are in the "monoids acting on
Applicatives" settings. It looks to me that in this case action on
domain probably is more natural.

**Comment by piyush-kurur** [comment]
I closed this thinking that there is no natural action on function
space that one can reasonably define. It was clear during the writing
of the paper that the action on type `b -> s` switches the
polarity. With this observation in mind I have the following natural
actions:

1. `m` action on `b` gives  `m` action on `something -> b`.

2. `m` action on `b` gives `Dual m` action on `b -> something`.

I feel this looks natural, does not need commutative and should be part
of the monoid-extras. What do you (@byorgey) think.

**Comment by byorgey** [comment]
That sounds reasonable, but the problem is that in Haskell we cannot have both those instances since they will overlap.  This is why we would need some `newtype`s.

**Comment by piyush-kurur** [comment]
I see your point. Closing this. Let me think a bit about this.
